src/transports/ipc.rs

```diff
@@ -0,0 +1,376 @@
+      .map(|ipc| (
+        EventLoopHandle { thread: Some(eloop), done: done },
+        ipc,
+      ))
+  }
+}
+
+impl Transport for Ipc {
+  type Out = BoxFuture<RpcValue, RpcError>;
+
+  fn execute(&self, method: &str, params: Vec<String>) -> Self::Out {
+    let id = self.id.fetch_add(1, atomic::Ordering::Relaxed);
+    let request = helpers::build_request(id, method, params);
+    debug!("Calling: {}", request);
+
+
+    // When the response is ready
+    let (tx, rx) = futures::oneshot();
+    let read_sender = self.read_sender.clone();
+    let write_sender = self.write_sender.clone();
+
+    read_sender.send((id, tx))
+      .then(|_| write_sender.send(request))
+      .then(|_| rx)
+      .then(|result| result.unwrap_or(Err(RpcError::Unreachable)))
+      .boxed()
+  }
+}
+
+enum WriteState {
+  WaitingForRequest,
+  Writing {
+    buffer: Vec<u8>,
+    current_pos: usize
+  },
+}
+
+/// Writing part of the IPC transport
+/// Awaits new requests using `mpsc::Receiver` and writes them to the socket.
+struct WriteStream {
+  write: WriteHalf<UnixStream>,
+  incoming: mpsc::Receiver<String>,
+  state: WriteState
+}
+
+impl Future for WriteStream {
+  type Item = ();
+  type Error = ();
+
+  fn poll(&mut self) -> futures::Poll<Self::Item, Self::Error> {
+    let new_state = match self.state {
+      WriteState::WaitingForRequest => {
+        // Ask for more to write
+        let to_send = try_ready!(self.incoming.poll());
+        if let Some(to_send) = to_send {
+          trace!("Got new message to write: {:?}", to_send);
+          WriteState::Writing {
+            buffer: to_send.into_bytes(),
+            current_pos: 0,
+          }
+        } else {
+          return Ok(futures::Async::NotReady);
+        }
+      },
+      WriteState::Writing { ref buffer, ref mut current_pos } => {
+        try_ready!(Ok(self.write.poll_write()));
+
+        // Write everything in the buffer
+        while *current_pos < buffer.len() {
+          let n = try_nb!(self.write.write(&buffer[*current_pos..]));
+          *current_pos += n;
+          if n == 0 {
+            warn!("Zero write.");
+            return Err(()); // zero write?
+          }
+        }
+
+        WriteState::WaitingForRequest
+      },
+    };
+
+    self.state = new_state;
+    self.poll()
+  }
+}
+/// Reading part of the IPC transport.
+/// Reads data on the socket and tries to dispatch it to awaiting requests.
+struct ReadStream {
+  read: ReadHalf<UnixStream>,
+  incoming: mpsc::Receiver<(usize, Pending)>,
+  pending: BTreeMap<usize, Pending>,
+  buffer: Vec<u8>,
+  current_pos: usize,
+}
+
+impl Future for ReadStream {
+  type Item = ();
+  type Error = ();
+
+  fn poll(&mut self) -> futures::Poll<Self::Item, Self::Error> {
+    // Check new incoming response channels
+    let response = self.incoming.poll();
+    if let Ok(futures::Async::Ready(Some((id, response)))) = response {
+      self.pending.insert(id, response);
+    }
```

Comment:
Right, there might be some responses pending in the channel. I will change it back to `Mutex`.